### DISCUSSION TITLE
Tarantool stat/info interface needs redesign

### DISCUSSION THREAD

**Discussion by Mons** [tarantool/tarantool#5833] [RFC]
# Tarantool stat/info interface needs redesign

Current contents of box.stat* is overflowed. Rather simple and compact in early versions it became almost unmaintainable and unreadable monster now. There are a lot of entry points for different kinds of information and I believe there is no person that could name them all without a doc or console.

What we have for now:
* `box.stat()` with box modifications(INSERT REPLACE UPSERT UPDATE DELETE), box queries (SELECT), iproto calls (EVAL CALL), weird ERROR (which counts... exceptions), SQL (PREPARE and EXECUTE) and low level AUTH
* `box.stat.net()` with similar to `box.stat()` interface and info about networking: connections, requests and bytes transmitted
* `box.stat.sql()` with strange format, not like in other `.stat()`'s
* `box.info.sql()`. contains info about caches of sql?
* `box.stat.vinyl()` with also different and hierarchical format (but worth a note: it's easy to read for a human). Better call it counters. and also it is accessible from `box.info.vinyl()`
* `box.info.memory()` — memory counters.
* `box.runtime.info()` - some kind of lua memory.
* `box.info.gc()` — information about _snapshot/xlog_ garbage collection. not lua gc.
* `box.info.election` — info about state of raft

There are also several entry points, that could be used to obtain information about internals, like `box.slab.info()`/`box.slab.stats()`, `fiber.info()`, `fiber.top()`, `require 'tarantool'` and maybe something I forget now. BTW: need for explicit `require` is very unfriendly for maintenance.

The documentation states

> The box.stat submodule provides access to request and network statistics.
> The box.info submodule provides access to information about server instance variables.

But this is far from reality.

And with this load if information we have a lack for:

* Iterator initializations (`index:pairs()`)/scans(`next()`)
* Separate counters for client errors, internal errors, iproto errors or `error()`'s
* Iproto level counters like authfail or schema mismatches
* Current executing requests and sql queries
* Length of internal queues, like queue between iproto and tx or applier queue or wal queue.

Also there were several discussions:
* We had a discussion about box.stat.ERRORS (https://github.com/tarantool/doc/issues/851)
* We had an old improvement request (https://github.com/tarantool/tarantool/issues/1031)

So, I have propositions:

1. We should separate points of view: `human` (admin or developer) or `machine` (internal or external metrics or monitoring)
    - human is lazy. expected input from human should be compact (no ugly options like `{ human = true }`)
    - tab-completion **must** work for human input
    - human entry point should be one entry point, that could be unfold when needed
    - human output **must** be easily readable
    - machine output may be in any form
2. We should separate the purpose: state analysis by humans or periodical fetch by metrics
    - metrics use accumulated counters (like box.stat.X.total) or absolute value (like box.stat.net.CONNECTIONS.current)
    - humans use "now" values, that are somehow prepared: (like box.stat.X.rps) or quota percents
3. We should separate info by subsystems:
    - runtime and app
      + luajit
      + eventloop
      + fibers
      + channels
    - storage
      + memtx
      + vinyl
    - storage access
      + ddl — data definition language (schema modifications)
      + dql — data query language (select, pairs)
      + dml — data modification language
    - network access
      + iproto
        * transport
        * protocol
      + hypothetical future protocol info
        * transport
        * protocol
    - replication
      + async
      + synchro
    - cluster
      + raft/membership
    - summary
        + load
        + memory
        + network
4. Summary statistics and counters should be aggregating from other layers. For example:
    - summary.memory should aggregate memory consumption from lua, mallocs, memtx and vinyl. Then it will be a guide for a user, that encounter issue with memory
    - summary.load may aggregate overall rps from different layers and cpu consumptions of main subsystems, so in overview it may show a bottleneck

THe proposition is to create another global entry point for maintenance (not `box.`). 
- It should be preloaded (no explicit require, or hack it within console)
- It should be used for instance management and gathering information

I propose namespace `srv.`.

Then, differentiate entry points for human and for machines by simple optional arguments:
    - `srv.info` — for humans
    - `srv.info({ format='yaml' | 'json' | 'rows' | ... })` — for machines

Use the following hierarchy:

```
srv
    info
        summary
            status
                TODO:
                    overall health status
                    if raft then membership status (voter|candidate|...)
                    writable
                    instance status ready | orphan | ...
                    errors, if any. from status, from replication, ...
                    uptime
            load
                TODO: cpu consumption
            memory
                TODO: memory consumption by subsystems
            network
                TODO: networking: connections, messages, bytes, rps
            storage
                TODO: storage usage in % or in volume
        build
            TODO: info from `require 'tarantool'.build`
        runtime
            luajit
                vm
                    TODO: something like misc.getstat
                gc
                    TODO: gc "pressure", object counts, gc step timings
            ev
                TODO: loop timings, epoll watcher stats
            fiber
                count
                TODO: sched info: summary csw's, creation/deletion
            channel
                count
                messages
        [?] memory
            TODO: small statistics
        [?] disk
            TODO: iops, traffic, iotime iowait
        storage
            memtx
                TODO: slab stats. kinda box.slab.stats, but readable
                TODO: space/index count/size
                TODO: averages (tuple size, so on...)
                TODO: fragmentation/waste
            vinyl
                TODO: kinda box.stat.vinyl
            wal
                TODO: wal
        net
            iproto
                TODO: transport level: packets/bytes. derive from box.stat.net()
                TODO: protocol level: requests. derive from box.stat()
            relay
                TODO: count of relays, traffic, tx count/s, ...
            applier
                TODO: count of appliers, traffic, tx count/s, ...
        stat
            ddl (data definition language (schema modifications))
                TODO: requests counts
            dql (data query language (select, pairs))
                TODO: requests counts
            dml (data modification language)
                TODO: requests counts
            sql
                TODO: derive from box.stat.sql
        replication
            TODO: derive from box.info.replication
            TODO: separate async / synchro
            TODO: xlog lock status (derive from box.info.gc)
        cluster
            election
                TODO: box.info.election
            raft/membership
        checkpoint
            TODO: box.info.gc
```

A lot of tickets related to this:
* [Display the number of active fibers somewhere](https://github.com/tarantool/tarantool/issues/1891)
* [Performance counters](https://github.com/tarantool/tarantool/issues/1989)
* [box.stat.admin](https://github.com/tarantool/tarantool/issues/1031)
* [Request statistic by masters](https://github.com/tarantool/tarantool/issues/2034)
* [Debug of requests](https://github.com/tarantool/tarantool/issues/3155)
* [Account SQL statements in box.stat](https://github.com/tarantool/tarantool/issues/4024)
* [IPROTO memory information](https://github.com/tarantool/tarantool/issues/933)
* [Display the number of active fibers somewhere](https://github.com/tarantool/tarantool/issues/1891)
* [Extend fiber.info() with total fiber CPU time](https://github.com/tarantool/tarantool/issues/2864)
* [fiber.info() details](https://github.com/tarantool/tarantool/issues/3190)
* [Include fiber creation stack in fiber.info()](https://github.com/tarantool/tarantool/issues/4002)
* [Make all built-in modules visible by default in the interactive console](https://github.com/tarantool/tarantool/issues/5596)
* [Improve box.slab.stats()](https://github.com/tarantool/tarantool/issues/5519)
* [Add box.info.replication.X.downstream.lag](https://github.com/tarantool/tarantool/issues/5447)
* [Add system version to build signature](https://github.com/tarantool/tarantool/issues/866)
* [https://github.com/tarantool/tarantool/issues/31](https://github.com/tarantool/tarantool/issues/31)

Also it's worth looking into closed tickets:
* [Memory usage overview](https://github.com/tarantool/tarantool/issues/934)
* [Rename values in box.slab.info() report](https://github.com/tarantool/tarantool/issues/1883) Here are very useful formats description

Keywords used to search issues: `box.info`, `box.stat`, `fiber.info`, `slab`

**Comment by no1seman** [comment]
Seems that it will be good if box will have a method like box.save_info() (or smth like this) which will save whole info from different places into one plaintext/yaml/json and also a tool that can analyze whole data from it and give some recomendations, for example: too many connections - tune sysctl or too big tuples - increase memtx_min_tuple_size and so on.
This report should not contain confidential information so that we can collect information from our customers and collect a knowledge base about real installations and there settings.

**Comment by filonenko-mikhail** [comment]
Also

```
net
   iproto
      connection
        *
          readahead buffer
              total
              usage
```

**Reply by Mons** [reply]
Yeah, those buffers causes readaheadache :)

**Comment by filonenko-mikhail** [comment]
I want to see all of this counters just from iproto thread without tx.
When we want to understand what's going wrong, tx is overloaded and cant help us.

**Comment by EvgenyMekhanik** [comment]
It would be nice to be able to build histograms for some types of statistics, for example, for the maximum length of the stream queue .

**Comment by filonenko-mikhail** [comment]
Please add coio thread pool monitoring

**Comment by filonenko-mikhail** [comment]
## The Things I Want To Feel

### Access to srv.info from anywhere (not only tx thread)

There ara cases when I (as a chief incident officer) want to know about Tarantool:

- How it is crashed, what is area of reasons
- Tarantool running, why latency are increasing right now
- Tarantool running without any traffic changes, why hardware load is increasing right now

Or other words 
- What profile of traffic is
- What Tarantool state is
- What OS state is


### The Cases 

There are incidents when user do not understand what's wrong. They are from different sides of Tarantool.

I want to describe cases in separate sections.

I want that we understand every case using srv.info for crashed and for running Tarantool

**Comment by filonenko-mikhail** [comment]
## Userstory The Readaheache Case

User wants to suppress readahead limit reached and this way increase perfomance.
User increases box.cfg.readahead.
User makes a lot of iproto connections.
Tarantool starts to use much memory.

- server.lua
```lua
local ffi = require("ffi")

local errno = require('errno')

ffi.cdef[[
typedef uint64_t rlim_t;

typedef struct rlimit {
    rlim_t rlim_cur;  /* Soft limit */
    rlim_t rlim_max;  /* Hard limit (ceiling for rlim_cur) */
} rlimit;

int getrlimit(int resource, struct rlimit *rlim);
int setrlimit(int resource, const struct rlimit *rlim);
]]

local RLIMIT_CORE = 4
local RLIMIT_NOFILE

if ffi.os == 'Linux' then
    RLIMIT_NOFILE = 7
elseif ffi.os == 'OSX' then
    RLIMIT_NOFILE = 8
end

local RLIM_INFINITY = bit.lshift(1ULL, 63) - 1
local OPEN_MAX = 10240

local function setrlimit(resource, limit)
    local rlimit = ffi.new('rlimit')
    rlimit.rlim_cur = limit.rlim_cur
    rlimit.rlim_max = limit.rlim_max
    local rc = ffi.C.setrlimit(resource, rlimit)
    if rc ~= 0 then
        return nil, errno.strerror()
    end

    return {rlim_cur = rlimit.rlim_cur,
            rlim_max = rlimit.rlim_max}
end

setrlimit(RLIMIT_NOFILE, {rlim_cur=10240, rlim_max=10240})

box.cfg{listen=3301, readahead=1024*1024*20}

box.schema.user.grant('guest', 'super', nil, nil, {if_not_exists=true})

require('console').start() os.exit()
```

- client.lua

```lua
local netbox = require('net.box')
local ffi = require("ffi")
local errno = require('errno')

ffi.cdef[[
typedef uint64_t rlim_t;

typedef struct rlimit {
    rlim_t rlim_cur;  /* Soft limit */
    rlim_t rlim_max;  /* Hard limit (ceiling for rlim_cur) */
} rlimit;

int getrlimit(int resource, struct rlimit *rlim);
int setrlimit(int resource, const struct rlimit *rlim);
]]

local RLIMIT_CORE = 4
local RLIMIT_NOFILE

if ffi.os == 'Linux' then
    RLIMIT_NOFILE = 7
elseif ffi.os == 'OSX' then
    RLIMIT_NOFILE = 8
end

local RLIM_INFINITY = bit.lshift(1ULL, 63) - 1
local OPEN_MAX = 10240

local function setrlimit(resource, limit)
    local rlimit = ffi.new('rlimit')
    rlimit.rlim_cur = limit.rlim_cur
    rlimit.rlim_max = limit.rlim_max
    local rc = ffi.C.setrlimit(resource, rlimit)
    if rc ~= 0 then
        return nil, errno.strerror()
    end

    return {rlim_cur = rlimit.rlim_cur,
            rlim_max = rlimit.rlim_max}
end

setrlimit(RLIMIT_NOFILE, {rlim_cur=10240, rlim_max=10240})

local clients = {}

for i = 1, 1e3 do 
    local c = netbox.connect('127.0.0.1:3301')
    c:eval([[ return 1 ]])
    table.insert(clients, c)
end

require('console').start() os.exit()
```

**Comment by filonenko-mikhail** [comment]
## The Infinite Loop

- server.lua
```
while true do
  the most useful stuf all over the world
end
```

**Comment by filonenko-mikhail** [comment]
## The Memory Usage

After OOM killer shooted, it's hard to know what subsystem is reason.
- readahead
- tuple arena
- memtx snapshot 
- vinyl
- wal
- relay
- somewhere else

**Comment by filonenko-mikhail** [comment]
## The Worst Fullscans

```lua
for i=1, #box.space.test:select() do
  box.space.other.get(i)
end
```

**Comment by no1seman** [comment]
Problem description:
In core stats we have: net.request.rps that collects all net_box requests and there is no ways to split exteranal requests and internal (intrcluster net_box requests). 
There's one way to go - wrap all callable functions with metrics wrapper, but seems it's ugly ans quite easy to forget to do it and also it's too much work to be done, also if adding some new stored function - have to wrap it. It's annoying.

Proposal:
Each instance exactly know who is who and who sent the request: one of cluster members or smbd else. What if we split net.request.rps into 2:
- net.request.rps_internal
- net.request.rps_external

One of the advantages of this metrics - we will easily get amplimification - a number of internal operations per one business request.

**Comment by filonenko-mikhail** [comment]
## The Fullscan Userstory

Environment:
- Sharded cluster
- CRUD
- External clients with complex read/write requests

Situation:
- Storage CPU 90%
- Too long wal write
- Request on router "Timeout exceeded"

Po konyam:
- fiber.info()
  - The things fine
  - No large csw values
- fiber.top()
  - One large average time **netbox** fiber
  - No stacktrace, no request body

Result:  
  After some time: the killer is select(nil)

I want to understand in fiber.top(), what request is load storage system:
  - **select** with space name, index name, index value, iterator
  - stored function name with args
  - peer ip

**Comment by filonenko-mikhail** [comment]
## The Сrashdump Userstory

Env:
- Cartridge Tarantool
- Requests to storage

Situation:
- Storage crashed
- GDB can't show backtrace
- GDB can't load thread local variables

- BUT GDB can load luajit.py and show tarantool_L

Result:
Killer still not found

I want to see lua backtraces without any tls magic. I have luajit.py and do not have LuaState for every fiber.

**Comment by filonenko-mikhail** [comment]
# Additional info about «applier in separate thread»
```
- num of applier threads
  - "*"
    - readahead buffer size
      - readahead used bytes
    - transaction queue to TX
```

**Comment by filonenko-mikhail** [comment]
## Crashdump Userstory 2

### Prerequisites

- There is crashdump
- Crashdump valid
- C stacktrace is good
- No Lua stacktrace because, no LuaState for current fiber.

## Problem

- Panic at yield in GC hook
- But c/lua stacktrace do not contain any __gc handler above
- This way we can't understand yielding gc hook

## Hypotesis

- Add stacktraces to all current fibers for all threads.

**Comment by ochaton** [comment]
It would be good to see number of transactions (box.commit calls). Or better commits and rollbacks of txn, because each transaction may contains many modifications

**Comment by filonenko-mikhail** [comment]
box.snapshot() can hang current caller fiber for a while. Its convenient to know who is caller of snapshot. Sometimes it is migration or ddl tools/scripts.

For e.g. etcd loggin message source every line.

**Comment by filonenko-mikhail** [comment]
These day there are several reason for instance to be readonly. Seems that it's convenient about one field: "reason instance to be read only": 
  - read_only = true
  - memory limit reached
  - replication_connect_quorum make orphan
  - raft failover
  - raft fencing
  - manual demote

**Reply by olegrok** [reply]
See box.info.ro_reason
https://github.com/tarantool/tarantool/commit/48185e8641fe83be16aa2b93d250ab0482f8e5cb

**Reply by filonenko-mikhail** [reply]
thx a lot, 2.10 become better and better
memory limit reached is absent ((

**Comment by filonenko-mikhail** [comment]
There is runtime arena other place for allocating tuples outside readahead and memtx and vinyl. Sometimes router with large workloads make runtime arena bigger and bigger.

**Comment by filonenko-mikhail** [comment]
incident

```
box.begin()

for tuple in index:pairs() do
  if some_buggy_condition then break end
  space.delete(tuple)
end

box.commit()
```
